### PR TITLE
feat(intelligence): add intelligent skill recommendations and creation

### PR DESCRIPTION
**Pull request by athola**
Introduce the `skrills-intelligence` crate providing smart skill recommendations based on usage patterns and project context analysis. This release establishes CLI parity for all intelligence MCP tools.

Key capabilities:
- Usage analytics from Claude Code and Codex CLI session data
- Project context analysis detecting languages, deps, and frameworks
- Smart recommendations combining dependency, usage, and context signals
- Skill creation via GitHub search or LLM generation

New CLI commands (mirroring MCP tools):
- resolve-dependencies: traverse skill dependency graphs
- recommend-skills-smart: intelligent multi-signal recommendations
- analyze-project-context: detect project languages and frameworks
- suggest-new-skills: identify gaps in skill coverage
- create-skill: generate skills via GitHub search or LLM
- search-skills-github: find existing skills on GitHub
- sync-from-claude alias for sync command

Documentation updates include expanded CLI reference, new ADRs (0003, 0004), and comprehensive book chapters covering the intelligence features.

### CODE REVIEW

**Review by athola — COMMENTED** [review]
## PR Review Summary

### ✅ Scope Compliance Analysis
**Source:** PR description and ADRs 0003/0004

**Requirements:**
1. ✅ Usage analytics from Claude Code and Codex CLI session data - Implemented
2. ✅ Project context analysis detecting languages, deps, and frameworks - Implemented  
3. ✅ Smart recommendations combining dependency, usage, and context signals - Implemented
4. ✅ Skill creation via GitHub search or LLM generation - Implemented
5. ✅ CLI parity for all intelligence MCP tools - Implemented
6. ✅ Documentation updates (ADRs, book chapters, architecture) - Implemented

### Code Quality Assessment
**Files Changed:** 59 files, +8,695/-491 lines
**Test Coverage:** All tests passing, good unit test coverage in new modules
**Clippy:** 0 warnings

### ✅ No Blocking Issues Found

The implementation is well-structured with:
- Clean separation of concerns (intelligence crate modularized into context/create/recommend/usage)
- Good test coverage for core functionality (scorer, analytics, parsers)
- Proper error handling throughout
- Thread-safe cache implementation with TTL and snapshot persistence
- Security-conscious GitHub API integration (token handling via env vars)

### In-Scope Observations (Non-blocking)

1. **[S1] Good architectural decision on crate versioning**
   - ADR-0004 documents independent versioning rationale well
   - Allows intelligence features to iterate faster without blocking other releases

2. **[S2] CLI parity well-executed**
   - All MCP tools have CLI equivalents per ADR-0003
   - Consistent --format text|json output patterns

3. **[S3] Recommendation scoring is configurable**
   - Weights are constants but could be exposed as configuration if needed later
   - Current design is simple and effective for v0.4.0

### Suggestions (Author's Discretion)

1. **[G1] Consider adding doc-tests to intelligence crate**
   - Location: crates/intelligence/src/lib.rs
   - Currently has 0 doc-tests; other crates have examples
   - Low priority - unit tests provide coverage

2. **[G2] Cache snapshot versioning**
   - Location: crates/server/src/cache.rs
   - If snapshot format changes in future, a version field could prevent stale loads
   - Current fingerprinting via roots is sufficient for now

### Quality Metrics
- **All tests passing**: ✅
- **Clippy warnings**: 0
- **Architecture**: Well-documented ADRs
- **Documentation**: Book chapters and CLI reference updated

### Recommendation
**APPROVED** ✅

This PR delivers comprehensive intelligence features with good code quality, proper documentation, and CLI parity. The modular architecture in the new intelligence crate sets a good foundation for future enhancements.

---
*Review generated by Claude Code PR Review*

**Comment by athola** [comment]
## PR Review Fixes Applied ✅

All critical and high-priority security issues from the code review have been addressed in commit 896b67f.

### Security Fixes
| Issue | Fix |
|-------|-----|
| Path traversal vulnerability | Added validation rejecting names with `..` or starting with `.` in `create_skill_tool` |
| GitHub query injection | Added `sanitize_github_query()` to strip search operators (`repo:`, `org:`, `stars:`, etc.) |
| CLI binary injection | Added `validate_cli_binary()` to reject shell metacharacters and path separators |

### Error Handling Improvements
| Issue | Fix |
|-------|-----|
| Silent failure in analytics loading | Converted `if let Ok(...)` to explicit `match` with `tracing::warn!` |
| Silent failures in context detection | Added `tracing::debug!` logging for README/git/dependency parsing failures |
| Unclear GitHub rate limit errors | Added actionable error messages for 401/403/422 responses |

### Test Coverage
- Added 7 tests for `sanitize_github_query()` covering operators, quoted values, boolean operators, and edge cases
- Added 4 tests for `validate_cli_binary()` covering path traversal, shell metacharacters, and edge cases
- All 303 workspace tests pass
- Clippy passes with no warnings

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by athola** [comment]
### Out-of-Scope Items → GitHub Issues

Following the `/fix-pr` Phase 3.5 workflow, the following GitHub issues have been created for deferred test coverage improvements:

| Review Item | Issue Created | Description |
|-------------|---------------|-------------|
| Low coverage (36.75%) | #55 | Integration tests for server intelligence tools |
| Low coverage (37-50%) | #56 | Tests for claude and codex usage parsers |
| Zero coverage (0%) | #57 | Tests for skill creation module |
| Low coverage (62.72%) | #58 | Integration tests for GitHub skill search |
| Low coverage (69.26%) | #59 | Integration tests for LLM skill generation |
| Low coverage (74-85%) | #60 | Tests for project context detection |
| Low coverage (61.17%) | #61 | Tests for skill recommendation scorer |

**Total Issues Created:** 7

These issues capture the medium/low priority items from the PR review that were out-of-scope for the immediate security fixes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: crates/intelligence/src/context/dependencies.rs [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+                                dev: true,
+                            });
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    // Parse [tool.poetry.dependencies] (Poetry)
```
Comment (athola) [inline crates/intelligence/src/context/dependencies.rs]: why only poetry? what about uv or other python package managers?

File: crates/intelligence/src/context/detector.rs [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+    }
+
+    frameworks.sort();
+    frameworks
+}
+
+/// Parse README file for description and keywords.
+fn parse_readme(root: &Path) -> Result<(Option<String>, Vec<String>)> {
+    let readme_names = [
+        "README.md",
```
Comment (athola) [inline crates/intelligence/src/context/detector.rs]: consider instead making this case insensitive when checking so you don't need to duplicate for upper/lowercase

File: crates/intelligence/src/create/cli_detector.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+pub fn get_cli_binary(env: CliEnvironment) -> &'static str {
+    match env {
+        CliEnvironment::ClaudeCode => "claude",
+        CliEnvironment::CodexCli => "codex",
+        CliEnvironment::Unknown => "claude", // Default to claude
+    }
+}
+
+/// Check if a CLI binary is available in PATH.
+#[allow(dead_code)]
```
Comment (athola) [inline crates/intelligence/src/create/cli_detector.rs]: add support for everything marked dead code. if support doesn't make sense, remove the functionality. dead code is a code smell

File: crates/intelligence/src/create/cli_detector.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        .arg(binary)
+        .output()
+        .map(|o| o.status.success())
+        .unwrap_or(false)
+}
+
+/// Get the best available CLI binary.
+#[allow(dead_code)]
+pub fn get_available_cli() -> Option<&'static str> {
+    let env = detect_cli_environment();
```
Comment (athola) [inline crates/intelligence/src/create/cli_detector.rs]: add support for everything marked dead code. if support doesn't make sense, remove the functionality. dead code is a code smell

File: crates/intelligence/src/usage/analytics.rs [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            let mut pairs: Vec<_> = coocs.iter().map(|(k, v)| (k.clone(), *v)).collect();
+            pairs.sort_by(|a, b| b.1.cmp(&a.1));
+            pairs.truncate(limit);
+            pairs
+        })
+        .unwrap_or_default()
+}
+
+/// Calculate a recency score (0.0 - 1.0) based on last usage.
+#[allow(dead_code)]
```
Comment (athola) [inline crates/intelligence/src/usage/analytics.rs]: add support for everything marked dead code. if support doesn't make sense, remove the functionality. dead code is a code smell

File: crates/intelligence/src/usage/claude_parser.rs [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+                }
+            }
+        }
+    }
+
+    Ok(events)
+}
+
+/// Parse Claude Code command history from ~/.claude/history.jsonl
+#[allow(dead_code)]
```
Comment (athola) [inline crates/intelligence/src/usage/claude_parser.rs]: add support for everything marked dead code. if support doesn't make sense, remove the functionality. dead code is a code smell

File: crates/intelligence/src/usage/codex_parser.rs [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+                Ok(session_events) => events.extend(session_events),
+                Err(e) => debug!("Failed to parse Codex session file {:?}: {}", path, e),
+            }
+        }
+    }
+
+    Ok(events)
+}
+
+#[allow(dead_code)]
```
Comment (athola) [inline crates/intelligence/src/usage/codex_parser.rs]: add support for everything marked dead code. if support doesn't make sense, remove the functionality. dead code is a code smell

File: crates/intelligence/src/context/detector.rs [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+                        }
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    // Find the language with the most files
+    let max_count = counts.values().map(|(c, _)| *c).max().unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/intelligence/src/context/detector.rs]: Fixed in commit 3c71f6c - Added `max_languages` parameter (defaults to 10, capped at 50) and exposed it through CLI `--max-languages` option and MCP tool parameter.

File: crates/intelligence/src/create/github_search.rs [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+/// Build a raw.githubusercontent.com URL for fetching file content.
+fn build_raw_url(full_name: &str, path: &str) -> String {
+    format!(
+        "https://raw.githubusercontent.com/{}/HEAD/{}",
+        full_name, path
+    )
+}
+
+/// Fetch the content of a skill from its raw URL.
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/intelligence/src/create/github_search.rs]: Fixed in commit 2e19953 - Exported `fetch_skill_content` and `search_skills_advanced` as part of the crate's public API.

File: crates/intelligence/src/create/github_search.rs [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+
+    if !response.status().is_success() {
+        anyhow::bail!("Failed to fetch skill content: {}", response.status());
+    }
+
+    Ok(response.text().await?)
+}
+
+/// Search for skills with specific criteria.
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/intelligence/src/create/github_search.rs]: Fixed in commit 2e19953 - See previous reply.

File: crates/intelligence/src/recommend/explainer.rs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+        }
+    } else if age_minutes > 0 {
+        format!("Used {} minutes ago", age_minutes)
+    } else {
+        "Just used".to_string()
+    }
+}
+
+/// Generate a summary of multiple recommendations.
+#[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/intelligence/src/recommend/explainer.rs]: Fixed in commit 2e19953 - Exported `summarize_recommendations` as part of the crate's public API.

File: crates/intelligence/src/usage/claude_parser.rs (hunk 6 above)
Comment (athola) [reply crates/intelligence/src/usage/claude_parser.rs]: Fixed in commit 2e19953 - Exported `parse_claude_command_history` as part of the crate's public API.

File: crates/server/src/commands/resolve.rs [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+    };
+
+    let direction_label = match (direction, transitive) {
+        (DependencyDirection::Dependencies, true) => "transitive dependencies",
+        (DependencyDirection::Dependencies, false) => "direct dependencies",
+        (DependencyDirection::Dependents, true) => "transitive dependents",
+        (DependencyDirection::Dependents, false) => "direct dependents",
+    };
+
+    if format == "json" {
```
[the remark this answers is not in the document]
Comment (athola) [reply crates/server/src/commands/resolve.rs]: Fixed in commit 3c71f6c - Created `OutputFormat` enum with `clap::ValueEnum` derive. Converted `--format` CLI args and matching logic to use the type-safe enum instead of strings.

**Comment by athola** [comment]
# Testing Plan: PR #54 - Intelligent Skills v0.4.0

This testing plan covers the new intelligent skill recommendations and creation features added in the `intelligent-skills-0.4.0` branch.

## Quick Start

```bash
# Build and install locally
cargo build --release
cp target/release/skrills ~/.codex/bin/skrills

# Run all automated tests first
cargo test --workspace
```

---

## 1. Automated Tests (CI)

### 1.1 Unit Tests

```bash
# Run all unit tests
cargo test --workspace

# Run intelligence-specific tests
cargo test -p skrills-intelligence

# Run server command tests
cargo test -p skrills-server commands::intelligence
```

**Expected:** All tests pass

### 1.2 Integration Tests

```bash
# Run all integration tests
cargo test --workspace -- --test-threads=1

# Specifically test the sync features
cargo test --test sync_integration
```

---

## 2. CLI Command Testing

Test each new command directly via the CLI.

### 2.1 `analyze-project-context`

**Purpose:** Detect project languages, frameworks, and dependencies.

```bash
# Test in current directory
skrills analyze-project-context

# Test with explicit directory
skrills analyze-project-context --project-dir ~/projects/my-rust-project

# Test with JSON output
skrills analyze-project-context --format json

# Test with git analysis disabled
skrills analyze-project-context --no-git

# Test with custom commit limit
skrills analyze-project-context --commit-limit 10
```

**Expected Output:**
- Languages detected with file counts
- Frameworks identified
- Dependencies listed by ecosystem
- Git keywords extracted (if enabled)

### 2.2 `recommend-skills-smart`

**Purpose:** Get intelligent skill recommendations using multiple signals.

```bash
# Basic recommendation (uses project context + usage)
skrills recommend-skills-smart

# Recommend based on a specific skill URI
skrills recommend-skills-smart --uri "skill://skrills/codex/rust-best-practices"

# Recommend based on prompt/query
skrills recommend-skills-smart --prompt "testing and debugging rust code"

# Limit results
skrills recommend-skills-smart --limit 5

# Disable usage analysis (faster)
skrills recommend-skills-smart --no-usage

# Disable context analysis
skrills recommend-skills-smart --no-context

# JSON output
skrills recommend-skills-smart --format json
```

**Expected Output:**
- Scored recommendations with explanations
- Signal breakdown (dependency, usage, context, prompt match)
- Score breakdown (if JSON)

### 2.3 `suggest-new-skills`

**Purpose:** Identify gaps in skill coverage based on project needs.

```bash
# Basic gap analysis
skrills suggest-new-skills

# With specific focus areas
skrills suggest-new-skills --focus-areas testing,security,deployment

# JSON output
skrills suggest-new-skills --format json

# Specific project directory
skrills suggest-new-skills --project-dir ~/projects/my-python-project
```

**Expected Output:**
- Skill gaps with priority levels
- Evidence for each gap
- Suggested skill names

### 2.4 `resolve-dependencies`

**Purpose:** Traverse skill dependency graphs.

```bash
# Resolve dependencies of a skill
skrills resolve-dependencies --uri "skill://skrills/codex/rust-best-practices"

# Resolve dependents (skills that depend on this one)
skrills resolve-dependencies --uri "skill://skrills/codex/testing" --dependents

# Limit depth
skrills resolve-dependencies --uri "skill://..." --depth 2

# JSON output
skrills resolve-dependencies --format json --uri "skill://..."
```

**Expected Output:**
- List of dependent/dependency URIs
- Proper depth traversal

### 2.5 `search-skills-github`

**Purpose:** Search GitHub for existing SKILL.md files.

```bash
# Search for rust-related skills
skrills search-skills-github --query "rust testing"

# Limit results
skrills search-skills-github --query "python" --limit 5

# JSON output
skrills search-skills-github --query "typescript" --format json
```

**Expected Output:**
- List of matching SKILL.md files on GitHub
- Repository info, stars, skill paths

**Note:** Requires `GITHUB_TOKEN` environment variable for API access.

### 2.6 `create-skill`

**Purpose:** Create new skills via GitHub search or LLM generation.

```bash
# Dry run - preview without creating
skrills create-skill --name "rust-error-handling" \
  --description "Best practices for error handling in Rust applications" \
  --dry-run

# Create via GitHub search only
skrills create-skill --name "python-testing" \
  --description "Python testing patterns and pytest usage" \
  --method github --dry-run

# Create via LLM generation only
skrills create-skill --name "custom-workflow" \
  --description "A workflow for my specific needs" \
  --method llm --dry-run

# Create with custom target directory
skrills create-skill --name "my-skill" \
  --description "Custom skill" \
  --target-dir ~/.claude/skills \
  --dry-run

# Create with project context
skrills create-skill --name "project-specific" \
  --description "Skill tailored to current project" \
  --project-dir . \
  --dry-run
```

**Expected Output:**
- GitHub search results (if method includes github)
- LLM-generated content (if method includes llm)
- Path to created skill (if not dry-run)

**Security Tests:**
```bash
# Should REJECT path traversal
skrills create-skill --name "../../../etc/passwd" --description "test" --dry-run
# Expected: Error about path traversal

# Should REJECT hidden files
skrills create-skill --name ".hidden-skill" --description "test" --dry-run
# Expected: Error about hidden files
```

---

## 3. MCP Tool Testing via Claude Code

Start a Claude Code session and test the MCP tools interactively.

### 3.1 Setup Verification

```
# In Claude Code session, verify MCP is loaded
/mcp

# Should show 'skrills' server with tools including:
# - recommend-skills-smart
# - analyze-project-context
# - suggest-new-skills
# - create-skill
# - search-skills-github
```

### 3.2 Test MCP Tools

**Test 1: Analyze Project Context**
```
Use the skrills MCP to analyze the current project context
```

**Test 2: Smart Recommendations**
```
Use skrills recommend-skills-smart to find skills relevant to this Rust project
```

**Test 3: Skill Gap Analysis**
```
Use skrills suggest-new-skills with focus on testing and deployment
```

**Test 4: GitHub Search**
```
Use skrills search-skills-github to find skills about "claude code automation"
```

**Test 5: Create Skill (Dry Run)**
```
Use skrills create-skill to preview creating a skill called "my-workflow"
with description "Automated workflow for common tasks" using dry_run mode
```

---

## 4. MCP Tool Testing via Codex CLI

Start a Codex session and test the same tools.

### 4.1 Setup Verification

```bash
# Start Codex
codex

# Verify MCP server is registered
# Check for skrills in the output
```

### 4.2 Test MCP Tools

The same tests from section 3.2 should work in Codex:

```
Analyze this project using the skrills MCP analyze-project-context tool
```

```
Get smart skill recommendations for this project
```

---

## 5. Integration Scenarios

### 5.1 Full Workflow: New Project Setup

1. Create a new test project:
   ```bash
   mkdir ~/test-project && cd ~/test-project
   cargo init --lib
   echo 'tokio = "1.0"' >> Cargo.toml
   ```

2. Analyze context:
   ```bash
   skrills analyze-project-context
   ```
   **Expected:** Detects Rust, tokio dependency

3. Get recommendations:
   ```bash
   skrills recommend-skills-smart
   ```
   **Expected:** Rust-related skills recommended

4. Find gaps:
   ```bash
   skrills suggest-new-skills --focus-areas testing
   ```
   **Expected:** Testing-related gaps identified

5. Clean up:
   ```bash
   rm -rf ~/test-project
   ```

### 5.2 Cross-Platform Parity

Test that CLI and MCP return equivalent results:

```bash
# CLI
skrills analyze-project-context --format json > /tmp/cli-context.json

# MCP (via Claude Code)
# Use analyze-project-context tool, compare structured output
```

### 5.3 Usage Analytics

Test that usage patterns are correctly parsed:

1. Use some skills in Claude Code session
2. Run:
   ```bash
   skrills recommend-skills-smart --format json
   ```
3. Verify usage signals appear in recommendations

---

## 6. Edge Cases and Error Handling

### 6.1 Invalid Inputs

```bash
# Invalid skill URI
skrills recommend-skills-smart --uri "not-a-valid-uri"

# Non-existent directory
skrills analyze-project-context --project-dir /nonexistent

# Empty query
skrills search-skills-github --query ""
```

### 6.2 Missing Dependencies

```bash
# Missing GITHUB_TOKEN
unset GITHUB_TOKEN
skrills search-skills-github --query "test"
# Expected: Graceful error or empty results

# Missing LLM configuration
skrills create-skill --name "test" --description "test" --method llm --dry-run
# Expected: Graceful error about LLM not available
```

### 6.3 Permission Issues

```bash
# Read-only directory
skrills create-skill --name "test" --description "test" \
  --target-dir /usr/share
# Expected: Permission error (not dry-run)
```

---

## 7. Performance Testing

### 7.1 Large Skill Library

```bash
# Time recommendation with many skills
time skrills recommend-skills-smart --format json > /dev/null
```

**Acceptable:** < 2 seconds for 100+ skills

### 7.2 Large Project Analysis

```bash
# Test on a large codebase
time skrills analyze-project-context --project-dir ~/large-project
```

**Acceptable:** < 5 seconds for typical projects

---

## 8. Security Validation

### 8.1 Path Traversal Prevention

```bash
# All should fail with security error
skrills create-skill --name "../escape" --description "test" --dry-run
skrills create-skill --name "foo/../bar" --description "test" --dry-run
skrills create-skill --name "foo/../../bar" --description "test" --dry-run
```

### 8.2 Hidden File Prevention

```bash
# Should fail
skrills create-skill --name ".hidden" --description "test" --dry-run
skrills create-skill --name ".." --description "test" --dry-run
```

### 8.3 Input Sanitization

Verify that skill names and descriptions don't allow code injection when creating skills.

---

## 9. Documentation Verification

Verify the CLI help text matches functionality:

```bash
skrills recommend-skills-smart --help
skrills analyze-project-context --help
skrills suggest-new-skills --help
skrills create-skill --help
skrills search-skills-github --help
skrills resolve-dependencies --help
```

---

## 10. Regression Testing

Ensure existing functionality still works:

```bash
# Core sync commands
skrills sync --help
skrills mirror --help
skrills validate --help
skrills analyze --help
skrills metrics --help

# Doctor command
skrills doctor
```

---

## Test Matrix Summary

| Test Category | CLI | MCP (Claude) | MCP (Codex) |
|--------------|-----|--------------|-------------|
| analyze-project-context | ✅ | ✅ | ✅ |
| recommend-skills-smart | ✅ | ✅ | ✅ |
| suggest-new-skills | ✅ | ✅ | ✅ |
| resolve-dependencies | ✅ | ✅ | ✅ |
| search-skills-github | ✅ | ✅ | ✅ |
| create-skill | ✅ | ✅ | ✅ |
| Security (path traversal) | ✅ | - | - |
| Error handling | ✅ | ✅ | ✅ |

---

## Acceptance Criteria

- [ ] All automated tests pass (`cargo test --workspace`)
- [ ] All 6 new CLI commands work with expected output
- [ ] MCP tools accessible from both Claude Code and Codex CLI
- [ ] Security validations prevent path traversal and hidden files
- [ ] Error handling is graceful (no panics, clear messages)
- [ ] Performance is acceptable (< 2s recommendations, < 5s analysis)
- [ ] Documentation/help text is accurate
- [ ] Existing sync/mirror functionality unaffected
